crypto-core/src/test/java/com/palantir/crypto2/jmh/EncryptionBenchmark.java

```diff
@@ -55,54 +60,107 @@
+            @Override
+            void writeTo(byte[] input, OutputStream output) throws IOException {
+                for (int i = 0; i < input.length; i += BUFFER_SIZE) {
+                    output.write(input, i, Math.min(BUFFER_SIZE, input.length - i));
+                }
+            }
+        };
+
+        abstract void writeTo(byte[] input, OutputStream output) throws IOException;
+    }
+
     @Benchmark
-    public final void gcmEncrypt(State state) throws NoSuchPaddingException, NoSuchAlgorithmException {
+    public final byte[] gcmEncrypt(State state) throws NoSuchPaddingException, NoSuchAlgorithmException {
         Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
         GCMParameterSpec gcmSpec = new GCMParameterSpec(8 * 16, state.key.getIv());
 
-        encrypt(state.data, cipher, state.key.getSecretKey(), gcmSpec);
+        return encrypt(state.writeStrategy, state.data, cipher, state.key.getSecretKey(), gcmSpec);
     }
 
     @Benchmark
-    public final void ctrEncrypt(State state) throws NoSuchPaddingException, NoSuchAlgorithmException {
+    public final byte[] ctrEncrypt(State state) throws NoSuchPaddingException, NoSuchAlgorithmException {
         Cipher cipher = Cipher.getInstance("AES/CTR/NoPadding");
         IvParameterSpec ivSpec = new IvParameterSpec(state.key.getIv());
-
-        encrypt(state.data, cipher, state.key.getSecretKey(), ivSpec);
+        return encrypt(state.writeStrategy, state.data, cipher, state.key.getSecretKey(), ivSpec);
     }
 
     @Benchmark
-    public final void apacheEncrypt(State state) throws IOException {
+    public final byte[] apacheEncrypt(State state) throws IOException {
         Properties props = ApacheCiphers.forceOpenSsl(new Properties());
 
+        // TODO(ckozak): implement BlackholeOutputStream wrapper around jmh Blackhole rather than buffering
         ByteArrayOutputStream baos = new ByteArrayOutputStream();
-        CtrCryptoOutputStream output = new CtrCryptoOutputStream(
-                props, baos, state.key.getSecretKey().getEncoded(), state.key.getIv());
-
-        output.write(state.data);
+        try (CtrCryptoOutputStream output = new CtrCryptoOutputStream(
+                props, baos, state.key.getSecretKey().getEncoded(), state.key.getIv())) {
+            state.writeStrategy.writeTo(state.data, output);
+        }
+        return baos.toByteArray();
     }
 
-    private void encrypt(byte[] bytes, Cipher cipher, Key key, AlgorithmParameterSpec spec) {
+    private byte[] encrypt(
+            WriteStrategy writeStrategy,
+            byte[] bytes,
+            Cipher cipher,
+            Key key,
+            AlgorithmParameterSpec spec) {
         try {
             cipher.init(Cipher.ENCRYPT_MODE, key, spec);
+            // TODO(ckozak): implement BlackholeOutputStream wrapper around jmh Blackhole rather than buffering
             ByteArrayOutputStream baos = new ByteArrayOutputStream();
```

Comment:
Either way I think we can pass in an output stream. In the benchmark case it can pass all args to the black hole, and in the validation case we can use baos and check the results.

Reply:
makes sense